File: core/ui/stt_popup.py
import sys
import json
import os
import random
import re
import zmq
from PyQt5.QtCore import Qt, QTimer, QPropertyAnimation, QEasingCurve, QPoint, QRect, QParallelAnimationGroup, QThread, pyqtSignal
from PyQt5.QtWidgets import QWidget, QLabel, QVBoxLayout, QHBoxLayout, QApplication, QGraphicsDropShadowEffect, QFrame, QSizePolicy
from PyQt5.QtGui import QFont, QColor, QFontDatabase, QPainter

class SttZmqListener(QThread):
    """Background ZMQ Subscriber Thread for STT UI updates"""
    status_received = pyqtSignal(dict)
    
    def run(self):
        context = zmq.Context()
        socket = context.socket(zmq.SUB)
        socket.connect("tcp://127.0.0.1:5556")
        
        socket.setsockopt_string(zmq.SUBSCRIBE, "STT_UPDATE")
        
        while True:
            try:
                message = socket.recv_string()
                topic, json_data = message.split(" ", 1)
                data = json.loads(json_data)
                self.status_received.emit(data)
            except Exception as e:
                pass

class WaveformIndicator(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setFixedSize(28, 18)
        self.bars = [3, 3, 3, 3] 
        self.target_bars = [3, 3, 3, 3]
        self.is_animating = False
        self.color = QColor(10, 132, 255)

        self.anim_timer = QTimer(self)
        self.anim_timer.timeout.connect(self.update_bars)
        
    def start_animation(self, color_hex):
        self.color = QColor(color_hex)
        self.is_animating = True
        self.anim_timer.start(40)
        
    def stop_animation(self, color_hex):
        self.color = QColor(color_hex)
        self.is_animating = False
        self.target_bars = [4, 4, 4, 4] 
        
    def update_bars(self):
        if self.is_animating:
            for i in range(4):
                if abs(self.bars[i] - self.target_bars[i]) < 1:
                    self.target_bars[i] = random.randint(3, 16)
        
        needs_update = False
        for i in range(4):
            diff = self.target_bars[i] - self.bars[i]
            if abs(diff) > 0.5:
                self.bars[i] += diff * 0.35 
                needs_update = True
                
        if needs_update or self.is_animating:
            self.update()

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.Antialiasing)
        painter.setPen(Qt.NoPen)
        painter.setBrush(self.color)
        
        bar_width = 3
        spacing = 3
        start_x = (self.width() - (4 * bar_width + 3 * spacing)) // 2
        
        for i, height in enumerate(self.bars):
            x = start_x + i * (bar_width + spacing)
            y = (self.height() - height) / 2
            painter.drawRoundedRect(int(x), int(y), bar_width, int(height), 1, 1)


class STTPopup(QWidget):
    def __init__(self):
        super().__init__()
        
        self.font_eng = QFont("Segoe UI", 13)
        eng_id = QFontDatabase.addApplicationFont("Data/fonts/english.ttf")
        if eng_id != -1:
            fam = QFontDatabase.applicationFontFamilies(eng_id)
            if fam: self.font_eng = QFont(fam[0], 13)

        self.font_hin = QFont("Nirmala UI", 14)
        hin_id = QFontDatabase.addApplicationFont("Data/fonts/devangri.ttf")
        if hin_id != -1:
            fam = QFontDatabase.applicationFontFamilies(hin_id)
            if fam: self.font_hin = QFont(fam[0], 14)

        self.current_state = "idle"
        self.target_geometry = None
        self.resize_anim = QPropertyAnimation(self, b"geometry")
        self.resize_anim.setEasingCurve(QEasingCurve.OutExpo)
        self.resize_anim.setDuration(200)

        self.initUI()
        
        self.transcribed_timer = QTimer()
        self.transcribed_timer.setSingleShot(True)
        self.transcribed_timer.timeout.connect(self.allow_hide)
        self.can_hide = True 
        self.last_text = ""

        self.zmq_listener = SttZmqListener()
        self.zmq_listener.status_received.connect(self.process_status_update)
        self.zmq_listener.start()
        
        self.process_status_update({"status": "idle", "text": ""})

    def initUI(self):
        self.setWindowFlags(Qt.FramelessWindowHint | Qt.WindowStaysOnTopHint | Qt.Tool | Qt.WindowTransparentForInput)
        self.setAttribute(Qt.WA_TranslucentBackground, True)
        self.setMinimumSize(1, 1)

        self.outer_layout = QVBoxLayout(self)
        self.outer_layout.setContentsMargins(40, 40, 40, 40)

        self.island = QFrame(self)
        self.island.setObjectName("Island")
        self.island.setStyleSheet("""
            #Island {
                background-color: #000000;
                border: 1px solid rgba(255, 255, 255, 0.15);
                border-radius: 24px;
            }
        """)
        self.island.setFixedHeight(48)

        self.shadow = QGraphicsDropShadowEffect(self)
        self.shadow.setBlurRadius(30)
        self.shadow.setColor(QColor(0, 0, 0, 150))
        self.shadow.setOffset(0, 4) 
        self.island.setGraphicsEffect(self.shadow)

        self.layout = QHBoxLayout(self.island)
        self.layout.setContentsMargins(22, 0, 22, 0) 
        self.layout.setSpacing(14)
        
        self.waveform = WaveformIndicator(self.island)
        self.waveform.setSizePolicy(QSizePolicy.Fixed, QSizePolicy.Fixed)
        
        self.text_label = QLabel("Listening...")
        self.text_label.setWordWrap(False)
        self.text_label.setSizePolicy(QSizePolicy.MinimumExpanding, QSizePolicy.MinimumExpanding)
        self.text_label.setAlignment(Qt.AlignVCenter | Qt.AlignLeft)
        
        self.set_fast_text_style("Listening...", "rgba(255, 255, 255, 0.5)", QFont.Normal)
        
        self.layout.addWidget(self.waveform, 0, Qt.AlignVCenter | Qt.AlignLeft)
        self.layout.addWidget(self.text_label, 1, Qt.AlignVCenter | Qt.AlignLeft)
        
        self.outer_layout.addWidget(self.island)
        self.setWindowOpacity(0.0)
        self.hide()

    def is_hindi(self, text):
        return bool(re.search(r'[\u0900-\u097F]', text))

    def set_fast_text_style(self, text, color, weight):
        font = self.font_hin if self.is_hindi(text) else self.font_eng
        font.setWeight(weight)
        self.text_label.setFont(font)
        self.text_label.setStyleSheet(f"color: {color}; background: transparent; border: none;")
        self.text_label.setText(text)

    def reset_island_style(self):
        self.island.setStyleSheet("""
            #Island {
                background-color: #000000;
                border: 1px solid rgba(255, 255, 255, 0.15);
                border-radius: 24px;
            }
        """)

    def calculate_target_geometry(self):
        self.text_label.adjustSize()
        
        island_width = self.text_label.sizeHint().width() + self.waveform.width() + 60
        ideal_width = max(200, min(island_width, 800)) + 80
        ideal_height = 48 + 80
        
        screen = QApplication.primaryScreen().availableGeometry()
        x = (screen.width() - ideal_width) // 2
        y = screen.bottom() - ideal_height - 50 
        
        return QRect(int(x), int(y), int(ideal_width), int(ideal_height))

    def smooth_resize(self):
        new_geometry = self.calculate_target_geometry()
        if self.target_geometry != new_geometry:
            self.target_geometry = new_geometry
            if self.isVisible() and self.windowOpacity() > 0.8:
                if self.resize_anim.state() == QPropertyAnimation.Running:
                    self.resize_anim.stop()
                self.resize_anim.setStartValue(self.geometry())
                self.resize_anim.setEndValue(new_geometry)
                self.resize_anim.start()
            else:
                self.setGeometry(new_geometry)

    def show_panel(self):
        if hasattr(self, 'hide_anim_group') and self.hide_anim_group.state() == QPropertyAnimation.Running:
            self.hide_anim_group.stop()

        self.smooth_resize()
        
        if not self.isVisible() or self.windowOpacity() < 1.0:
            start_y = self.target_geometry.y() + 40 
            self.setGeometry(self.target_geometry.x(), start_y, self.target_geometry.width(), self.target_geometry.height())
            self.show()
            self.raise_()

            self.show_anim_group = QParallelAnimationGroup(self)
            fade_in = QPropertyAnimation(self, b"windowOpacity")
            fade_in.setDuration(200)
            fade_in.setStartValue(self.windowOpacity()) 
            fade_in.setEndValue(1.0)
            
            slide_up = QPropertyAnimation(self, b"pos")
            slide_up.setDuration(350) 
            slide_up.setStartValue(self.pos())
            slide_up.setEndValue(QPoint(self.target_geometry.x(), self.target_geometry.y()))
            slide_up.setEasingCurve(QEasingCurve.OutBack)

            self.show_anim_group.addAnimation(fade_in)
            self.show_anim_group.addAnimation(slide_up)
            self.show_anim_group.start()

    def hide_panel(self):
        if not self.isVisible(): return
        if hasattr(self, 'show_anim_group') and self.show_anim_group.state() == QPropertyAnimation.Running:
            self.show_anim_group.stop()

        self.hide_anim_group = QParallelAnimationGroup(self)
        fade_out = QPropertyAnimation(self, b"windowOpacity")
        fade_out.setDuration(200)
        fade_out.setStartValue(self.windowOpacity())
        fade_out.setEndValue(0.0)
        
        slide_down = QPropertyAnimation(self, b"pos")
        slide_down.setDuration(250)
        slide_down.setStartValue(self.pos())
        slide_down.setEndValue(QPoint(self.x(), self.y() + 20))
        slide_down.setEasingCurve(QEasingCurve.InBack)

        self.hide_anim_group.addAnimation(fade_out)
        self.hide_anim_group.addAnimation(slide_down)
        self.hide_anim_group.finished.connect(self.hide)
        self.hide_anim_group.start()

    def allow_hide(self):
        self.can_hide = True
        self.process_status_update({"status": self.current_state, "text": self.last_text})

    def process_status_update(self, data):
        self.current_state = data.get("status", "idle")
        text = data.get("text", "")

        if self.current_state == "exit":
            QApplication.quit()
            return

        if self.current_state == "idle":
            self.reset_island_style() 
            self.waveform.stop_animation("#444444") 
            if not self.can_hide:
                return 
            self.hide_panel()
            self.last_text = ""
            return
            
        if self.current_state in ["listening", "understanding"]:
            self.reset_island_style()
            self.waveform.start_animation("#0A84FF")
            
            if not text.strip() and self.current_state == "listening":
                self.set_fast_text_style("Listening...", "rgba(255, 255, 255, 0.4)", QFont.Normal)
            else:
                display_text = text if text.strip() else self.last_text
                self.last_text = display_text.capitalize() if not self.is_hindi(display_text) else display_text
                self.set_fast_text_style(self.last_text, "rgba(255, 255, 255, 0.95)", QFont.Medium)
            
        elif self.current_state == "transcribed":
            self.can_hide = False
            self.transcribed_timer.start(2000) 
            self.waveform.stop_animation("#32D74B")
            
            self.island.setStyleSheet("""
                #Island {
                    background-color: #000000;
                    border: 1px solid rgba(50, 215, 75, 0.5);
                    border-radius: 24px;
                }
            """)
            
            display_text = text.capitalize() if text.strip() and not self.is_hindi(text) else (text if text.strip() else self.last_text)
            self.set_fast_text_style(display_text, "#FFFFFF", QFont.Normal)

        self.show_panel()

if __name__ == "__main__":
    app = QApplication(sys.argv)
    panel = STTPopup()
    sys.exit(app.exec_())

File: core/UiSrc/SttPopup/Stt_Popup.cpp
#include "stt_popup.h"
#include <QRandomGenerator>

// ==========================================
// WAVEFORM INDICATOR
// ==========================================
WaveformIndicator::WaveformIndicator(QWidget *parent) : QWidget(parent) {
    setFixedSize(28, 18);
    bars = {3.0, 3.0, 3.0, 3.0};
    target_bars = {3.0, 3.0, 3.0, 3.0};
    is_animating = false;
    color = QColor("#0A84FF");

    anim_timer = new QTimer(this);
    connect(anim_timer, &QTimer::timeout, this, &WaveformIndicator::update_bars);
}

void WaveformIndicator::start_animation(const QString &color_hex) {
    color = QColor(color_hex);
    is_animating = true;
    anim_timer->start(8); // 120 FPS (1000ms / 120 = ~8ms)
}

void WaveformIndicator::stop_animation(const QString &color_hex) {
    color = QColor(color_hex);
    is_animating = false;
    target_bars = {4.0, 4.0, 4.0, 4.0};
}

void WaveformIndicator::update_bars() {
    if (is_animating) {
        for (int i = 0; i < 4; ++i) {
            if (std::abs(bars[i] - target_bars[i]) < 1.0) {
                target_bars[i] = QRandomGenerator::global()->bounded(3, 17);
            }
        }
    }

    bool needs_update = false;
    for (int i = 0; i < 4; ++i) {
        double diff = target_bars[i] - bars[i];
        if (std::abs(diff) > 0.1) {
            bars[i] += diff * 0.1; // Smooth factor for 120fps
            needs_update = true;
        }
    }

    if (needs_update || is_animating) {
        update();
    }
}

void WaveformIndicator::paintEvent(QPaintEvent *event) {
    Q_UNUSED(event);
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing);
    painter.setPen(Qt::NoPen);
    painter.setBrush(color);

    int bar_width = 3;
    int spacing = 3;
    int start_x = (width() - (4 * bar_width + 3 * spacing)) / 2;

    for (int i = 0; i < 4; ++i) {
        int x = start_x + i * (bar_width + spacing);
        double height = bars[i];
        int y = (this->height() - height) / 2.0;
        painter.drawRoundedRect(x, y, bar_width, height, 1, 1);
    }
}

// ==========================================
// STT POPUP MAIN WINDOW (UDP LOGIC)
// ==========================================
STTPopup::STTPopup(QWidget *parent) : QWidget(parent), current_state("idle"), can_hide(true) {
    font_eng = QFont("Segoe UI", 13);
    // YAHAN PATH UPDATE KIYA HAI (:/ lagaya hai)
    int eng_id = QFontDatabase::addApplicationFont(":/Data/fonts/english.ttf");
    if (eng_id != -1) {
        QStringList fam = QFontDatabase::applicationFontFamilies(eng_id);
        if (!fam.isEmpty()) font_eng = QFont(fam.first(), 13);
    }

    font_hin = QFont("Nirmala UI", 14);
    // YAHAN PATH UPDATE KIYA HAI (:/ lagaya hai)
    int hin_id = QFontDatabase::addApplicationFont(":/Data/fonts/devangri.ttf");
    if (hin_id != -1) {
        QStringList fam = QFontDatabase::applicationFontFamilies(hin_id);
        if (!fam.isEmpty()) font_hin = QFont(fam.first(), 14);
    }

    resize_anim = new QPropertyAnimation(this, "geometry", this);
    resize_anim->setEasingCurve(QEasingCurve::OutExpo);
    resize_anim->setDuration(200);

    show_anim_group = new QParallelAnimationGroup(this);
    hide_anim_group = new QParallelAnimationGroup(this);

    transcribed_timer = new QTimer(this);
    transcribed_timer->setSingleShot(true);
    connect(transcribed_timer, &QTimer::timeout, this, &STTPopup::allow_hide);

    initUI();

    // ==========================================
    // UDP SOCKET INITIALIZATION
    // ==========================================
    udpSocket = new QUdpSocket(this);
    udpSocket->bind(QHostAddress::LocalHost, 5556); // Same port as Python
    connect(udpSocket, &QUdpSocket::readyRead, this, &STTPopup::readPendingDatagrams);

    QJsonObject initial_status;
    initial_status["status"] = "idle";
    initial_status["text"] = "";
    process_status_update(initial_status);
}

void STTPopup::readPendingDatagrams() {
    while (udpSocket->hasPendingDatagrams()) {
        QNetworkDatagram datagram = udpSocket->receiveDatagram();
        QByteArray data = datagram.data();

        QJsonDocument doc = QJsonDocument::fromJson(data);
        if (!doc.isNull() && doc.isObject()) {
            process_status_update(doc.object());
        }
    }
}

void STTPopup::initUI() {
    setWindowFlags(Qt::FramelessWindowHint | Qt::WindowStaysOnTopHint | Qt::Tool | Qt::WindowTransparentForInput);
    setAttribute(Qt::WA_TranslucentBackground, true);
    setMinimumSize(1, 1);

    outer_layout = new QVBoxLayout(this);
    outer_layout->setContentsMargins(40, 40, 40, 40);

    island = new QFrame(this);
    island->setObjectName("Island");
    reset_island_style();
    island->setFixedHeight(48);

    shadow = new QGraphicsDropShadowEffect(this);
    shadow->setBlurRadius(30);
    shadow->setColor(QColor(0, 0, 0, 150));
    shadow->setOffset(0, 4);
    island->setGraphicsEffect(shadow);

    layout = new QHBoxLayout(island);
    layout->setContentsMargins(22, 0, 22, 0);
    layout->setSpacing(14);

    waveform = new WaveformIndicator(island);
    waveform->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);

    text_label = new QLabel("Listening...", island);
    text_label->setWordWrap(false);
    text_label->setSizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::MinimumExpanding);
    text_label->setAlignment(Qt::AlignVCenter | Qt::AlignLeft);

    set_fast_text_style("Listening...", "rgba(255, 255, 255, 0.5)", QFont::Normal);

    layout->addWidget(waveform, 0, Qt::AlignVCenter | Qt::AlignLeft);
    layout->addWidget(text_label, 1, Qt::AlignVCenter | Qt::AlignLeft);

    outer_layout->addWidget(island);
    setWindowOpacity(0.0);
    hide();
}

bool STTPopup::is_hindi(const QString &text) {
    QRegularExpression rx("[\\x{0900}-\\x{097F}]");
    return rx.match(text).hasMatch();
}

void STTPopup::set_fast_text_style(const QString &text, const QString &color, QFont::Weight weight) {
    QFont font = is_hindi(text) ? font_hin : font_eng;
    font.setWeight(weight);
    text_label->setFont(font);
    text_label->setStyleSheet(QString("color: %1; background: transparent; border: none;").arg(color));
    text_label->setText(text);
}

void STTPopup::reset_island_style() {
    island->setStyleSheet(
        "#Island {"
        "   background-color: #000000;"
        "   border: 1px solid rgba(255, 255, 255, 0.15);"
        "   border-radius: 24px;"
        "}"
        );
}

QRect STTPopup::calculate_target_geometry() {
    text_label->adjustSize();

    int island_width = text_label->sizeHint().width() + waveform->width() + 60;
    int ideal_width = qMax(200, qMin(island_width, 800)) + 80;
    int ideal_height = 48 + 80;

    QRect screen = QApplication::primaryScreen()->availableGeometry();
    int x = (screen.width() - ideal_width) / 2;
    int y = screen.bottom() - ideal_height - 50;

    return QRect(x, y, ideal_width, ideal_height);
}

void STTPopup::smooth_resize() {
    QRect new_geometry = calculate_target_geometry();
    if (target_geometry != new_geometry) {
        target_geometry = new_geometry;
        if (isVisible() && windowOpacity() > 0.8) {
            if (resize_anim->state() == QPropertyAnimation::Running) {
                resize_anim->stop();
            }
            resize_anim->setStartValue(geometry());
            resize_anim->setEndValue(new_geometry);
            resize_anim->start();
        } else {
            setGeometry(new_geometry);
        }
    }
}

void STTPopup::show_panel() {
    if (hide_anim_group->state() == QPropertyAnimation::Running) {
        hide_anim_group->stop();
    }

    smooth_resize();

    if (!isVisible() || windowOpacity() < 1.0) {
        int start_y = target_geometry.y() + 40;
        setGeometry(target_geometry.x(), start_y, target_geometry.width(), target_geometry.height());
        show();
        raise();

        show_anim_group->clear();
        QPropertyAnimation *fade_in = new QPropertyAnimation(this, "windowOpacity");
        fade_in->setDuration(200);
        fade_in->setStartValue(windowOpacity());
        fade_in->setEndValue(1.0);

        QPropertyAnimation *slide_up = new QPropertyAnimation(this, "pos");
        slide_up->setDuration(350);
        slide_up->setStartValue(pos());
        slide_up->setEndValue(QPoint(target_geometry.x(), target_geometry.y()));
        slide_up->setEasingCurve(QEasingCurve::OutBack);

        show_anim_group->addAnimation(fade_in);
        show_anim_group->addAnimation(slide_up);
        show_anim_group->start();
    }
}

void STTPopup::hide_panel() {
    if (!isVisible()) return;
    if (show_anim_group->state() == QPropertyAnimation::Running) {
        show_anim_group->stop();
    }

    hide_anim_group->clear();
    QPropertyAnimation *fade_out = new QPropertyAnimation(this, "windowOpacity");
    fade_out->setDuration(200);
    fade_out->setStartValue(windowOpacity());
    fade_out->setEndValue(0.0);

    QPropertyAnimation *slide_down = new QPropertyAnimation(this, "pos");
    slide_down->setDuration(250);
    slide_down->setStartValue(pos());
    slide_down->setEndValue(QPoint(x(), y() + 20));
    slide_down->setEasingCurve(QEasingCurve::InBack);

    hide_anim_group->addAnimation(fade_out);
    hide_anim_group->addAnimation(slide_down);

    disconnect(hide_anim_group, &QParallelAnimationGroup::finished, this, &STTPopup::hide_panel_finished);
    connect(hide_anim_group, &QParallelAnimationGroup::finished, this, &STTPopup::hide_panel_finished);

    hide_anim_group->start();
}

void STTPopup::hide_panel_finished() {
    hide();
}

void STTPopup::allow_hide() {
    can_hide = true;
    QJsonObject data;
    data["status"] = current_state;
    data["text"] = last_text;
    process_status_update(data);
}

void STTPopup::process_status_update(QJsonObject data) {
    current_state = data.value("status").toString("idle");
    QString text = data.value("text").toString("");

    if (current_state == "exit") {
        QApplication::quit();
        return;
    }

    if (current_state == "idle") {
        reset_island_style();
        waveform->stop_animation("#444444");
        if (!can_hide) return;
        hide_panel();
        last_text = "";
        return;
    }

    if (current_state == "listening" || current_state == "understanding") {
        reset_island_style();
        waveform->start_animation("#0A84FF");

        if (text.trimmed().isEmpty() && current_state == "listening") {
            set_fast_text_style("Listening...", "rgba(255, 255, 255, 0.4)", QFont::Normal);
        } else {
            QString display_text = text.trimmed().isEmpty() ? last_text : text;
            if (!is_hindi(display_text) && !display_text.isEmpty()) {
                display_text[0] = display_text[0].toUpper();
            }
            last_text = display_text;
            set_fast_text_style(last_text, "rgba(255, 255, 255, 0.95)", QFont::Medium);
        }
    } else if (current_state == "transcribed") {
        can_hide = false;
        transcribed_timer->start(2000);
        waveform->stop_animation("#32D74B");

        island->setStyleSheet(
            "#Island {"
            "   background-color: #000000;"
            "   border: 1px solid rgba(50, 215, 75, 0.5);"
            "   border-radius: 24px;"
            "}"
            );

        QString display_text = text.trimmed().isEmpty() ? last_text : text;
        if (!display_text.isEmpty() && !is_hindi(text)) {
            display_text[0] = display_text[0].toUpper();
        }
        set_fast_text_style(display_text, "#FFFFFF", QFont::Normal);
    }

    show_panel();
}